modules/cart/src/Plugin/Block/CartBlock.php

```diff
@@ -109,11 +110,18 @@ public function blockSubmit($form, FormStateInterface $form_state) {
    *   A render array.
    */
   public function build() {
+    $cachable_metadata = new CacheableMetadata();
+    $cachable_metadata->addCacheContexts(['user', 'session']);
+
     /** @var \Drupal\commerce_order\Entity\OrderInterface[] $carts */
     $carts = $this->cartProvider->getCarts();
     $carts = array_filter($carts, function ($cart) {
       /** @var \Drupal\commerce_order\Entity\OrderInterface $cart */
-      return $cart->hasLineItems();
+      // There is a chance the cart may have converted from a draft order, but
+      // is still in session. Such as just completing check out. So we verify
+      // that the cart is still a cart.
+      // @todo find better place to put this, such as provider::isFinalized
```

Comment:
Well it's half true. We can't remove it from the session, but we have no real way of identifying the order in session is no longer a cart. It's an edge case on anonymous users and checkout and that we have one session variable key for all carts, instead of like 1.x which had two.

But changing it would take a lot of work and should possibly be just a follow-up, since this block is the one bug experience.